Implement several more world methods

simon816 opened the pull request:
I have tested all of these methods except `setSeed` and `setBiome`

There are some comments which it'd be good to have some input on

Minecrell commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -93,21 +125,106 @@ public BlockLoc getBlock(int x, int y, int z) {
+            if (spawnEntityInWorld(entity)) {
+                return Optional.of((Entity) entity);
+            }
+        } catch (ReflectiveOperationException ignored) {
```
This doesn't compile, see https://travis-ci.org/SpongePowered/Sponge/builds/45614519#L514-L517

simon816 replied:
Oh, `ReflectiveOperationException` is Java 1.7, I'll change it

Deamon5550 commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -93,21 +125,106 @@ public BlockLoc getBlock(int x, int y, int z) {
+    public Optional<Entity> getEntityFromUUID(UUID uuid) {
+        for (net.minecraft.entity.Entity entity : this.loadedEntityList) {
+            if (entity.getUniqueID().equals(uuid)) {
+                return Optional.of((Entity) entity);
```
There is a better way to do this, WorldServer holds a Map of 'entitiesByUuid' which you can use to directly lookup the entity.

simon816 replied:
@Deamon5550 Done

simon816 commented:
I have updated this PR to add a few more methods.
The subproject was updated in order to get the changes to the biome API, and so I have also fixed the WorldBorder mixin to change radius => border.

There are a few comments on some lines, I invite you to add github comments so I can make the most appropriate changes

AIOOB commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -105,21 +138,54 @@ public BlockLoc getBlock(int x, int y, int z) {
+            // Entity constructor does not have single World parameter, too bad
+        }
         return Optional.absent();
     }
```
It's probably safer to use `EntityList.createEntityByName(String, World)`

simon816 replied:
I suppose it might be, but since we already know the class (`type.getEntityClass()`) it's probably quicker to use that than to use a map lookup

bloodmc commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -260,4 +328,94 @@ public boolean doesKeepSpawnLoaded() {
+    public boolean deleteChunk(Chunk chunk) {
+        IChunkProvider provider = this.getChunkProvider();
+        Vector3i chunkPos = chunk.getPosition();
+        if (provider instanceof ChunkProviderServer) {
```
Every worldserver creates a ChunkProviderServer instance, there is no need to check anything else. There also needs to be a check for forced chunks as many forge mods contain blocks that keep chunks loaded. If a forced chunk is found, this method needs to return false.

bloodmc commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -260,4 +328,94 @@ public boolean doesKeepSpawnLoaded() {
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public Iterable<Chunk> getLoadedChunks() {
```
Same issue as below, this method only needs to return the loaded chunks of the ChunkProviderServer.

progwml6 commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -260,4 +328,94 @@ public boolean doesKeepSpawnLoaded() {
+
+    @Override
+    public void setSeed(long seed) {
+        ObfuscationReflectionHelper.setPrivateValue(WorldInfo.class, this.worldInfo, seed, "randomSeed");
```
@bloodmc do we want to use AT's for things like this to prevent errors during MC version updates?

bloodmc replied:
Yes, this should be added to our sponge_at.cfg

bloodmc commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -126,21 +159,56 @@ public BlockLoc getBlock(int x, int y, int z) {
     @Override
     public void setBiome(Vector3i position, BiomeType biome) {
-        // TODO
+        net.minecraft.world.chunk.Chunk chunk = this.getChunkFromBlockCoords(VecHelper.toBlockPos(position));
```
This needs a check for null.

Zidane commented on src/main/java/org/spongepowered/mod/mixin/world/MixinWorld.java:
```diff
@@ -288,4 +361,88 @@ public void setKeepSpawnLoaded(boolean keepLoaded) {
+    @Override
+    public Map<String, String> getGameRules() {
+        GameRules gameRules = this.worldInfo.getGameRulesInstance();
+        // TODO Could expose GameRules#theGameRules via reflection to remove the
```
Just make this public in the AT...

simon816 replied:
OK, so there's no problem with exposing the underlying Map I take it.

simon816 replied:
Actually, there's not really much advantage of doing that so I'll remove the comment.
